0012-common-ledger-api/0012-common-ledger-api.md

```diff
@@ -0,0 +1,1201 @@
+
+**Authorization:** The `from` field of the message MUST match the account owner. Unauthenticated users MUST NOT be able to send messages. A ledger MAY allow admin connections to send messages whose `from` value matches the ledger's base [URL][].
+
+#### Request Format
+
+POST a JSON [Message resource][] to the `message` [URL][].
+
+**Metadata URL Name:** `message`. Suggested generic path: `message`
+
+#### Response Format
+
+A successful response uses the HTTP response code **204 No Content** and contains no message body.
+
+#### Example
+
+Request:
+
+```json
+POST /messages
+Content-Type: application/json
+
+{
+  "from": "https://blue.ilpdemo.ripple.com/ledger/accounts/bob",
+  "to": "https://blue.ilpdemo.ripple.com/ledger/accounts/alice",
+  "ledger": "https://blue.ilpdemo.ripple.com/ledger",
+  "data": {
+    "method": "quote_request",
+    "id": "721e4126-98a1-4974-b35a-8a8f4655f934",
+    "data": {
+      "source_amount": "100.25",
+      "source_address": "example.eur-ledger.alice",
+      "destination_address": "example.usd-ledger.bob",
+      "source_expiry_duration": "6000",
+      "destination_expiry_duration": "5"
+    }
+  }
+}
+```
+
+**Note:** This example uses the [Interledger Quoting Protocol](https://github.com/interledger/rfcs/blob/master/0008-interledger-quoting-protocol/0008-interledger-quoting-protocol.md) quote request type as example data. However, the `data` field may contain arbitrary objects, so the ledger MUST NOT require a specific format for such data.
+
+
+Response:
+
+```
+HTTP/1.1 204 No Content
+```
+
+#### Errors
+
+- [InvalidBodyError][]
+- [InvalidUriParameterError][]
+- [UnprocessableEntityError][]
+
+
+### Get Auth Token
+[Get Auth Token]: #get-auth-token
+
+Get a token that can be used to authenticate future requests.
+
+**Note:** This method is REQUIRED for ledgers to authenticate [WebSocket][] connections. If the ledger does not authenticate WebSocket connections, this method is OPTIONAL. The ledger MAY allow clients to authenticate for any other methods of the API using the tokens returned by this method.
+
+#### Request Format
+
+GET the `auth_token` [URL][].
+
+Depending on the [authentication mechanism](#authorization-and-authentication) used by this ledger, the ledger MAY require the HTTP `Auth` header with a valid username and password, or the ledger may require a different method of authentication.
+
+**Metadata URL Name:** `auth_token`.
+
+#### Response Format
+
+A successful response uses the HTTP response code **200 OK** and contains a JSON object with a token that can be used for subsequent requests. The ledger can generate the token with [JWT](https://jwt.io/) or any similar system.
+
+#### Example
+
+Request:
+
+```
+GET /auth_token
+Authorization: Basic myUsername:securePassphrase
+```
+
+Response:
+
+```json
+HTTP/1.1 200 OK
+Content-Type: application/json
+
+{
+  "token": "9AtVZPN3t49Kx07stO813UHXv6pcES"
+}
+```
+
+#### Errors
+
+- [Unauthorized][]
+
+
+
+
+## WebSocket
```

Comment:
I was playing with an alternative solution to websocket based on gRPC. First thing I needed was to abstract all the  "websocket subscription and message stuff" as a bidirectional asynchronous stream. 

An early *proto defining the service is available at:
https://github.com/everis-innolab/ILP_gRPC_API/blob/master/ilp_ledger_subscriptions.proto

A comparative about a different alternative solutions for bidirectional async channels:

\  | websocket+JSON/RPC 2.0 | Client request + Client pooling | protobuf + gRPC | web-hooks
-- | ---------------------- | ------------------------------- | -------------   | ---------
Notes| - Current five-bells implementation - Once the ws connection is stablished client and ledger become peers.|- Simplest solution.  - Messages from client to ledger are sent in HTTP requests.  Messages from ledger to client are pulled during periodic pooling requests.|- bidirectional async stream are created with: streamN(stream msg) returns (stream msg) |both ledger and client must provide a callback URL for peers messages/events
web-browser support|YES|YES| "WORK IN PROGRESS" github.com/grpc/grpc-experiments/issues/159 github.com/grpc-ecosystem/grpc-gateway |NO Android support |YES|YES|YES|YES
iOS support|YES|YES|YES|YES
Standard|YES (IETF RFC 6455)|NO|"Open-Source / IETF Draft"|NO
Works on 3G, behind proxies/firewalls,...| "SOMETIMES" |YES|YES/NO/SOMETIMES?| NO (both peers require IP visibility to expose REST web-hooks)
automatic mapping object to/from network data"|PARTIAL and IN SOME LANGUAJES|NO|YES|NO
auto-documented |NO|NO| YES. Using *proto files ( *proto can also be used as ref doc.  for non-gRPC implementations ) |NO
cross-language support |NO|NO|YES|NO
speed|GOOD|SLOW |VERY GOOD|UNDEFINED
"deprecated" support (backward compatibility)|MANUAL|MANUAL|YES|MANUAL
PREFERED FOR| CONNECTORS & SERVER APPS | SORT-LIVE CONNECTIONS WHERE MOBILE CLIENTS AND RANDOM NETWORK SETUPS ARE EXPECTED. (""WALLETS"") | CONNECTORS & SERVER APPS | CONNECTORS & SERVER APPS IF WEBSOCKETS AND gRPC ARE NOT AVAILABLE. 


I'm also thinking whether it could be convenient to split this API in services for light clients ("wallets" that just connect time to time to the ledger posting new prepared Transfers and checking payments), server applications ("e-comerce" with permanent tcp/ip connections waiting for payments and fulfilling transfer) and connectors. Also many simultaneous "wallet" connections to ledger could be expected in some setups while only a very limited set of ilp-connectors and server apps ("e-commerce"). If I'm not wrong there is no distinction amongst them right now. The API treats them all equally.